drivers/led_strip/tlc59731.c

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2024 Javad Rahimipetroudi <javad.rahimipetroudi@mind.be>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_tlc59731
+
+/**
+ * @file
+ * @brief LED driver for the TLC59731 LED driver.
+ *
+ * TLC59731 is a 3-Channel, 8-Bit, PWM LED Driver
+ * With Single-Wire Interface (EasySet)
+ *
+ * The EasySet protocol is based on short pulses and the time between
+ * them. At least one pulse must be sent every T_CYCLE, which can be
+ * between 1.67us and 50us. We want to go as fast as possible, but
+ * delays under 1us don't work very well, so we settle on 5us for the
+ * cycle time.
+ * A pulse must be high for at least 14ns. In practice, turning a GPIO on
+ * and immediately off again already takes longer than that, so no delay
+ * is needed there.
```

Comment:
I have some concerns about this implementation.
As @thedjnK has already commented, I think it is difficult to guarantee the timing.
As commented here, it may work without delay in the current environment, but it does not necessarily work on other boards.
This is undesirable since Zephyr drivers can be used regardless of board or SoC. (I think it's allowed to set it so it can't be used on a specific board, but it needs to be shown that it can't work as a design.)
In cases like this, Existing LED strip drivers mainly create waveforms using SPI or I2S. It's probably the most desirable.

Reply:
It's true that bitbanging on a GPIO risks messing up the timing. However, if the TLC59731 isn't connected to a SPI-MOSI or I2S pin, then bitbanging over GPIO is the only option.
So I think the driver should eventually get all three interfaces, with device tree properties defining which type of interface to use. But I think this should be taken step by step, and the GPIO version sounds like the simplest one to start with because it's the most generic one.

Reply:
I join @thedjnK  and @soburi here.

@javad123javad please consider an SPI implementation. I can see that the TLC59731 controller is connected to a SPI muxed pin on the STM32WB5MM-DK board (the board you are using). So let's use it.

I remember I already made this comment: https://github.com/zephyrproject-rtos/zephyr/pull/68455#issuecomment-1926295292.

Reply:
Should the GPIO implementation (i.e. the implementation that also works in cases where the LEDs are not connected to a MOSI or I2S capable pin) be removed completely then? Or should the correct one be chosen in the device tree?

In either case, how should it be modeled in the device tree? Should it be as a sub-node of the SPI master (even though it's not SPI at all)? If so, should the pinmux be dynamically adapted to avoid that the clock is output and potentially confuses other devices? Or should it be left to the device tree writer to disable the default pinmux for the SPI master and instead use something that only enables the MOSI pin? Or do you have a different solution in mind?

Is there an existing driver that abuses SPI to bitbang some other protocol?

> I think it is difficult to guarantee the timing.

The spi_bitbang driver and i2c_bitbang driver don't seem to share that concern, and AFAICS they don't do anything like disabling interrupts either. The spi_bitbang driver [makes the frequency configurable from device tree](https://elixir.bootlin.com/zephyr/latest/source/drivers/spi/spi_bitbang.c#L54), perhaps a similar property should be added?

Reply:
> Thanks for your comments, I'm working on the SPI implementation of the driver.

Thanks ! That's being said... Since you are almost done with the GPIO driver, I suggest you finish it. I am not sure it will be merged. There are pros and cons to be discussed. But even if it doesn't end up in the Zephyr tree, it is interesting to have this driver ready "somewhere", at least for debugging purpose. For example with the WS2812 and compatibles LED controllers, we have the `ws2812_gpio` driver. It is not suitable for production needs. But it is sometimes useful to have it around. Since you are close with the GPIO driver, please finish it. I'll review it this week-end.

Reply:
Hi @arnout,

> Should the GPIO implementation (i.e. the implementation that also works in cases where the LEDs are not connected to a MOSI or I2S capable pin) be removed completely then? Or should the correct one be chosen in the device tree?

I don't know. I think it is interesting to have the GPIO driver at least for debugging purposes. But it must be clear that it not adapted to production needs. Let's wait to hear from others.

> 
> In either case, how should it be modeled in the device tree? Should it be as a sub-node of the SPI master (even though it's not SPI at all)? If so, should the pinmux be dynamically adapted to avoid that the clock is output and potentially confuses other devices? Or should it be left to the device tree writer to disable the default pinmux for the SPI master and instead use something that only enables the MOSI pin? Or do you have a different solution in mind?
> 
> Is there an existing driver that abuses SPI to bitbang some other protocol?

Yes there is. See the `ws2812` LED controller. We have 4 compatibles for the same device: `worldsemi,ws2812-gpio`, `worldsemi,ws2812-i2s`, `worldsemi,ws2812-rpi_pico-pio`, `worldsemi,ws2812-spi`. We also have 4 drivers (one for each compatible) for the `ws2812`controller. And the `ws2812_spi` driver abuses the SPI protocol to output a signal comprehensible by stripped WS2812 controllers.

We could use the same method for the TLC59731.

OR we could follow this recommendation from @mbolivar-ampere: https://github.com/zephyrproject-rtos/zephyr/pull/55226#discussion_r1186525434.

This would allow to have a single compatible with several drivers. In the driver the [DT_ANY_INST_ON_BUS_STATUS_OKAY](https://docs.zephyrproject.org/latest/build/dts/api/api.html#c.DT_ANY_INST_ON_BUS_STATUS_OKAY) or [DT_INST_ON_BUS](https://docs.zephyrproject.org/latest/build/dts/api/api.html#c.DT_INST_ON_BUS) macros can be used to check the `on-bus` property. And in kconfig the [dt_compat_on_bus](https://docs.zephyrproject.org/latest/build/kconfig/preprocessor-functions.html) function can be used to enable SPI support if the `on-bus` property is set to `spi.

It is my preferred choice.

> 
> > I think it is difficult to guarantee the timing.
> 
> The spi_bitbang driver and i2c_bitbang driver don't seem to share that concern, and AFAICS they don't do anything like disabling interrupts either. The spi_bitbang driver [makes the frequency configurable from device tree](https://elixir.bootlin.com/zephyr/latest/source/drivers/spi/spi_bitbang.c#L54), perhaps a similar property should be added?

The SPI implementation should only be used in the case where a TLC59731 controller strip is connected to a SPI MOSI pin. And interrupts won't be an issue. Ideally DMA will be enabled and used. The idea for this driver is to use a SPI engine to output the correct waveform, offloading the CPU. So if no SPI engine can be used, then there is no point in using SPI protocol at all :) In this case the GPIO version of this driver should be preferred.

But in practice, when a such LED controller is built into a board, it is almost always wired to an SPI-muxed pin (as with the STM32WB5MM DK board). And in the case of an add-on module, then it's up to the user to connect it to a SPI MOSI pin on its board. So almost all the time you have a better option that GPIOs to drive a LED strip controller. That's why we are asking @javad123javad for a SPI implementation.

I hope this answers your questions :)

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/70763 is a workaround that can be used, depending upon the length of time it needs to be locked for determines if it should be used or not